Enable alpha on Block Inspector Color ToolsPanel

ramonjd opened the pull request:
Resolves #18095

## Description

This PR add alpha controls to Color Picker panels in the Block Inspector. 

We're doing this by passing enableAlpha=true to the underlying ColorPicker component from the color tools panel.

Where **text color** alpha transparency goes below a minimum threshold (currently 1), we'll also show a warning about readability where the text is otherwise "readable" according to the colord library.

Changes to background color transparency are ignored due to the difficultly in accounting for multiple alpha layers or image backgrounds and so on.

See: https://github.com/WordPress/gutenberg/issues/18095#issuecomment-1005377529


https://user-images.githubusercontent.com/6458278/150261226-82aa23ff-bec4-4965-8f47-02a08cd329a4.mp4


#### Considerations

Here we're enabling alpha controls for all blocks and all colors.

Is the availability to adjust alpha something we'd want to enable at a block level, e.g., via color supports and/or for certain colors, e.g., background, text, link?

Or is it okay to leave it on as a permanent citizen of the color panel?

## How has this been tested?
Insert any block that has color support, for example a group block!

Select a background color, or any color. Click on the custom color picker, then play with the alpha slider.

<img width="844" alt="Screen Shot 2022-01-18 at 10 49 03 am" src="https://user-images.githubusercontent.com/6458278/149848671-7f932f91-6137-4777-9687-5765e441f255.png">

The colors in the editor should match the frontend.

Take note of the warning.
<img width="273" alt="Screen Shot 2022-01-19 at 10 56 35 pm" src="https://user-images.githubusercontent.com/6458278/150125098-1bfcfd5e-dac1-46a6-b240-1017c4211e82.png">

The warning should show where **only the text color has transparency** AND the color panel does not show any contrast warnings otherwise.


## Types of changes
Enabling an existing feature.

## Checklist:
- [x] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ramonjd committed:
Initial commit: passing enableAlpha=true to the underlying ColorPicker component from the color tools panel.

ramonjd commented:
Testing on a few blocks

<img width="649" alt="Screen Shot 2022-01-18 at 11 13 32 am" src="https://user-images.githubusercontent.com/6458278/149850119-003540f5-185b-4170-b936-b4a93085e72a.png">

jasmussen commented:
Took this one for a quick spin, and it works as intended. Before:
<img width="1392" alt="before" src="https://user-images.githubusercontent.com/1204802/149893690-ce6406ad-06fe-43b8-bd31-8ea4e6f1f000.png">

After:

<img width="1392" alt="after" src="https://user-images.githubusercontent.com/1204802/149893714-481006cb-bd55-4045-8835-b841953d2ec6.png">

Honestly, this seems fine to me. The picker is designed to manage the prominence of both the hue and alpha selectors, and it can easily survive being visible forever.

Are there any reasons why we would _not_ want to surface alpha as an option? Are there any cases where changing the opacity of a color would actually not work? I can't think of any, can you @jameskoster?

Otherwise, this one seems fine to me! Nice work.

aristath commented:
Allowing alpha for backgrounds makes sense. However, I'm not so sure about the text & link colors. Using semi-transparent text would make the text a lot harder to read, and would be an a11y issue.
Granted, it's all about allowing users to make the choices they want, but at the same time we have to balance how much we're enabling bad choices.
I can imagine semi-transparent text being pretty when used in a gigantic header with a background image in a cover block for example, but that's an exception and not the norm. Text should not be allowed to be transparent IMO.

jasmussen commented:
> However, I'm not so sure about the text & link colors. Using semi-transparent text would make the text a lot harder to read, and would be an a11y issue.
> Granted, it's all about allowing users to make the choices they want, but at the same time we have to balance how much we're enabling bad choices.

Good thought. Do you think we could handle that with the same type of accessibility notice you receive if you choose badly contrasting colors? A semitransparent text color when blended into the background feels very comparable in that it's about comparing two colors for contrast.

If it's technically tricky to compare the contrast of semitransparent text on a background, we could just blanket output a warning as soon as the opacity isn't 100%. Someting like "Be careful with transparent text, contrast is likely going to be insufficient."

aristath commented:
> If it's technically tricky to compare the contrast of semitransparent text on a background, we could just blanket output a warning as soon as the opacity isn't 100%. Someting like "Be careful with transparent text, contrast is likely going to be insufficient."

Comparing the contrast for semi-transparent text would be extremely difficult... Theoretically, if alpha = 0.6 then we'd mix the text-color with the background-color with a ratio of 60-40 to get a hex value for the text color, which can then be compared against a hex for the background. But if the background itself has a transparency - or is an image, then any computation will be invalid.
A blanket warning like the one described above would make sense if opacity is used for text/link 👍

jameskoster commented:
Yeah, text is the only thing I can think of that might be tricky. A blanket warning seems like a sensible idea.

dennisheiden commented:
> Comparing the contrast for semi-transparent text would be extremely difficult... Theoretically, if alpha = 0.6 then we'd mix the text-color with the background-color with a ratio of 60-40 to get a hex value for the text color, which can then be compared against a hex for the background. But if the background itself has a transparency - or is an image, then any computation will be invalid.
> A blanket warning like the one described above would make sense if opacity is used for text/link 👍

Or you just check if a text block has a text color with an alpha lower than 1. You don't have to check against background's alphas for that to trigger an accessibility warning, because every text with an alpha can be considered as less readable than normal text.
At least in that way you can at least make people aware of the accessibility issue, otherwise they just create a style with alpha.

ramonjd commented:
Thanks for the feedback, folks. 

I lean towards offering the toolbox to allows users make their own decisions, so a blanket warning when the alpha for text moves beyond `n` threshold seems to strike a good balance.

I'll take a shot at it. 👍 

Later, we still might want to look at building alpha support into block supports, so we can control where the alpha control appears (e.g., only for backgrounds). 

In a block's block.json we could have something like (example only):

```json
"supports": {
    "color": {
       "background" : {
            "opacity" : true,
       },
       "text" : true,
    }
}
```
Whether alpha/opacity/whatever is disabled or enabled) by default is another question. 😄

ramonjd committed:
Show a warning in the color panel where either the text or background color has transparency.
Adding tests.
Updating README.md to reflect the reality that colord does not check the contrast of colors with alpha transparency in our implementation.

ramonjd commented on packages/block-editor/src/components/contrast-checker/index.js:
```diff
@@ -71,16 +83,19 @@ function ContrastChecker( {
 	const colordTextColor = colord( textColor || fallbackTextColor );
 	const hasTransparency =
 		colordBackgroundColor.alpha() !== 1 || colordTextColor.alpha() !== 1;
```
Should we reduce the minimum alpha threshold here where `__experimentalEnableAlphaChecker === true`? Say to `0.8`, so the warning will only show when the transparency of either text or background falls below this?

jasmussen commented:
Took it for a quick spin and seeing this:
![colors](https://user-images.githubusercontent.com/1204802/150129777-42b00866-8eb4-4ba7-a57e-31a1dcf574ec.gif)

It works well, although I think we should show this notice _only_ for text, and not for the background, because even with a transparent background you can get yourself into trouble with the wrong background image. 

How about this text: "Be careful with transparent text, contrast can often be insufficient."

On a separate note and not for this PR, I would like to update our notices across the board to be simpler. This is just an unfinished mockup at this point, but something along these lines:

<img width="364" alt="Paragraph w  contrast notice" src="https://user-images.githubusercontent.com/1204802/150132876-e8b20338-c681-4d64-b82b-9048b7ce2d74.png">

ramonjd commented:
> although I think we should show this notice only for text, and not for the background, because even with a transparent background you can get yourself into trouble with the wrong background image.

Good idea, thanks. This might reduce notice fatigue too. 😄

ramonjd committed:
enableContrastChecking if enableAlpha is true
only show the transparency warning where the text color alpha is < 1
ensure the readability/contrast warning takes precedence over the transparency warning

ramonjd commented:
> although I think we should show this notice only for text, and not for the background, because even with a transparent background you can get yourself into trouble with the wrong background image.

I've updated the code and the screencast in the test description.

> How about this text: "Be careful with transparent text, contrast can often be insufficient."

Removing the "for people" part would be good. It seems redundant. 

To my ear, something along the lines of "Transparent text may be hard to read." rings better.

We don't mention contrast sufficiency elsewhere I guess because it's leaning towards the technical too much? 🤷 

I dunno, I'm just shouting at the forest. 😄 

For now I've left it at `Transparent text colors may be hard for people to read.` to keep in line with current copy.

Also, I noticed the contrast checker doesn't take link colors into account. 


![Jan-20-2022 13-42-34](https://user-images.githubusercontent.com/6458278/150262677-c25f3194-079c-4966-9a3f-1d9290ffc657.gif)

I see there's an [issue](https://github.com/WordPress/gutenberg/issues/37549) up already. 

I think I have a way forward on this. I'll add another PR.

glendaviesnz commented:
This tested well for me. It will be interesting to see what this means over time for the likes of the Cover block explicit dimRatio setting - I suspect there might still be the need for some blocks to expose this option at a higher level still?

Is it likely to be changeable enough to warrant giving an `__experimental` prefix to `enableAlphaChecker` - given that this doesn't affect saved block content at all could we skip committing it as experimental?

ramonjd commented:
Thanks for testing @glendaviesnz 

> It will be interesting to see what this means over time for the likes of the Cover block explicit dimRatio setting

The Cover Block is an interesting one - it explicitly allows background images and the opacity layer. 

Though some sites out there do some rough calculations to guesstimate the readability of transparent backgrounds - see for example https://contrast-ratio.com/ which seems to be doing something [similar to what was described in this comment](https://github.com/WordPress/gutenberg/pull/37731#issuecomment-1015177694) - it's not accurate. 

The [color library](https://github.com/omgovich/colord) we use appears to flat out ignore alpha for testing readability and contrast too. Throw in background images, RGBA, opacity and unlimited layers of semi-transparent, nested blocks and 😱 

Ultimately, we could try to extend the colord library with a fudge and at least attempt to advise/warn users of potentially unreadable color mixes, but it runs the risk of become noisy.

> Is it likely to be changeable enough to warrant giving an __experimental prefix to enableAlphaChecker - given that this doesn't affect saved block content at all could we skip committing it as experimental?

That's a good question. I added the `__experimental` prefix right at the beginning before I knew how I was going to tackle things, mainly as a bit of butt-covering manoeuvre.

I think I agree that we can ditch it.

ramonjd committed:
Rename enable alpha prop to remove the experimental prefix

jasmussen commented:
<img width="297" alt="Screenshot 2022-01-20 at 10 06 02" src="https://user-images.githubusercontent.com/1204802/150307682-6f909edd-4a37-478d-b047-200cd7baf93d.png">

Looks good! Nice and simple.

ramonjd commented:
I appreciate the input, everyone! 🙇

uniquesolution commented:
I am not seeing this alpha option in block editor for group or other blocks that support color, using wordpress 5.9 RC3 and gutenberg 12.4.1, will this feature be available in the future release ? or I have to do some changes in theme.json to enable it ?

ramonjd commented:
> I am not seeing this alpha option in block editor for group or other blocks that support color, using wordpress 5.9 RC3 and gutenberg 12.4.1, will this feature be available in the future release ? or I have to do some changes in theme.json to enable it ?

HI @uniquesolution !

This change was merged a few days ago so, unless something catastrophic occurs like a worldwide internet failure or coffee shortages, it should be part of Gutenberg 12.5. 👍
